### PR TITLE
compat expansion; forge refactor

### PR DESCRIPTION
**Pull request by brianolson**
## Description
Expand "compat" forge test to simultaneously do traffic generation, gather stats, *and* run a gradual upgrade of validator nodes. Ensure that TPS stays high enough during upgrade.

Lots of refactor to get there, replacing lots of `&Foo` and `&mut Foo` with `Arc<Mutex<Foo>>` to make things multithread capable.

## Type of Change
- [X] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [X] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [X] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
This is tests. This PR running tests in forge clusters is the test.

## Key Areas to Review
Is this idiomatic Rust? Is there a better way?

## Checklist
- [X] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I identified and added all stakeholders and component owners affected by this change as reviewers
- [X] I tested both happy and unhappy path of the functionality
- [X] I have made corresponding changes to the documentation

### CODE REVIEW

File: crates/transaction-emitter-lib/src/emitter/submission_worker.rs
Comment (perryjrandall) [inline crates/transaction-emitter-lib/src/emitter/submission_worker.rs:261]: are these intentionally removed? should remove them?

Comment (brianolson) [reply crates/transaction-emitter-lib/src/emitter/submission_worker.rs]: Fixed. `update_seq_num_and_get_num_expired()` got refactored away because I decided it was two operations awkwardly joined together with awkward value passing boundaries, which I think got more awkward when some of the values changed type due to other refactoring here. All cleaned up now.

**Review by perryjrandall — APPROVED** [review]
This is a big change, my only questions are

1) what if any change do we expect in most tests? and how will we verify? (check forge stable after landing? check on pr time?) please be explicit in your description
2) here or later should we continue this to the logical extension? instead of mut and passing lifetimes should we just arc / mutex / rwlock everything so that its all fully parallelizeable and has a nicer interface?

Accepting to unblock, test is passing but would love to know what the next steps / direction here are? 

Make more of forge more parallel? Will this save time / cost / can we get better signal from existing tests?

**Comment by brianolson** [comment]
The most likely failure mode (which I encountered several times while getting the tests I have run to pass) is that some function declares a new tokio Runtime and calls `.block_on()` inside some function that is now newly under an `async` function anywhere in the call stack above it. This will panic and crash. It's annoying but 'not hard' to fix. The usual fix is to make the function with the Runtime to instead itself be `async` and then `block_on(foo())` -> `foo().await`.

I have **probably** missed some cases where this will happen, and there will probably be test failures that then take 30-60 minutes each of developer time to fix. I know I caught like a dozen sites where this would have happened; I probably missed some. If there's a tag we can apply to this PR to "run all forge the tests", we should do that?

This change creates a **better signal** from the `compat` test. I think being able to measure continued sustained TPS _during_ a rolling restart is a pretty good thing! We could apply that to other tests if they are currently serialized phases of "measure TPS", "do change", "measure TPS".

reference vs Arc? Yeah, there will probably be more of that. This much is just what I needed to get the compat change I wanted. It looks like a big change because it touched some interfaces that were propagated to many parts of the code. If you pretend that changing the signature of a function that's implemented 18 times is 'one change' then it's smaller? :-D

**Review by igor-aptos — COMMENTED** [review]
thanks for changing all these, having all these be async is so much cleaner

File: testsuite/testcases/src/consensus_reliability_tests.rs
Comment (brianolson) [reply testsuite/testcases/src/consensus_reliability_tests.rs:41]: to make sure I get exactly the scope I want on the read-lock object

**Comment by brianolson** [comment]
@igor-aptos found a bunch of things my brain had clearly glossed over after reading the diff too many times. Fixed those and some more cleanups I found from doing another fresh read through.

Changing the `compat` duration _might_ be significant. I'll be watching the current run with interest.

**Review by igor-aptos — APPROVED** [review]

**Pull request merged**
